plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/FileSystemExchangeSink.java

```diff
@@ -0,0 +1,373 @@
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+import io.trino.spi.exchange.ExchangeSink;
+import org.openjdk.jol.info.ClassLayout;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.crypto.SecretKey;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.net.URI;
+import java.util.ArrayDeque;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.MoreFutures.addExceptionCallback;
+import static io.airlift.concurrent.MoreFutures.addSuccessCallback;
+import static io.airlift.concurrent.MoreFutures.asVoid;
+import static io.airlift.concurrent.MoreFutures.toCompletableFuture;
+import static io.airlift.slice.SizeOf.estimatedSizeOf;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.concurrent.CompletableFuture.failedFuture;
+
+@ThreadSafe
+public class FileSystemExchangeSink
+        implements ExchangeSink
+{
+    public static final String COMMITTED_MARKER_FILE_NAME = "committed";
+    public static final String DATA_FILE_SUFFIX = ".data";
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FileSystemExchangeSink.class).instanceSize();
+
+    private final FileSystemExchangeStorage exchangeStorage;
+    private final URI outputDirectory;
+    private final int outputPartitionCount;
+    private final Optional<SecretKey> secretKey;
+    private final BufferPool bufferPool;
+
+    private final Map<Integer, BufferedStorageWriter> writersMap = new ConcurrentHashMap<>();
+    private final AtomicReference<Throwable> failure = new AtomicReference<>();
+    private volatile boolean closed;
+
+    public FileSystemExchangeSink(FileSystemExchangeStorage exchangeStorage, URI outputDirectory, int outputPartitionCount, Optional<SecretKey> secretKey)
+    {
+        this.exchangeStorage = requireNonNull(exchangeStorage, "exchangeStorage is null");
+        this.outputDirectory = requireNonNull(outputDirectory, "outputDirectory is null");
+        this.outputPartitionCount = outputPartitionCount;
+        this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        // double buffering to overlap computation and I/O
+        this.bufferPool = new BufferPool(outputPartitionCount * 2, exchangeStorage.getWriteBufferSize());
```

Comment:
I'm thinking whether it makes sense to have a minimum. For example when writing into a single partition (broadcast join, arbitrary exchange) having more buffering space might increase throughput. How about something like  `min(outputPartitionCount * 2, 10)` (or maybe even having a property). This can be a separate commit / separate PR though.